floors/fetcher.go

```diff
@@ -0,0 +1,323 @@
+}
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
+		return nil, openrtb_ext.FetchNone
+	}
+
+	// Check for floors JSON in cache
+	result, found := f.Get(config.Fetcher.URL)
+	if found {
+		var fetchedFloorData openrtb_ext.PriceFloorRules
+		err := json.Unmarshal(result, &fetchedFloorData)
+		if err != nil || fetchedFloorData.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return &fetchedFloorData, openrtb_ext.FetchSuccess
+	}
+
+	//miss: push to channel to fetch and return empty response
+	if config.Enabled && config.Fetcher.Enabled && config.Fetcher.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetcher, FetchTime: f.time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
+	floorData, fetchedMaxAge := f.fetchAndValidate(config)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", config.URL)
+		cacheExpiry := config.MaxAge
+		if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = fetchedMaxAge
+		}
+		floorData, err := json.Marshal(floorData)
+		if err != nil {
+			glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
+		} else {
+			f.SetWithExpiry(config.URL, floorData, cacheExpiry)
+		}
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
+
+}
+
+// Stop terminates fetcher service can closes all the channels and destroys worker pool
+func (f *PriceFloorFetcher) Stop() {
+	if f == nil {
+		return
+	}
+
+	close(f.done)
+	f.pool.Stop()
+	close(f.configReceiver)
+}
+
+func (f *PriceFloorFetcher) submit(fetchInfo *FetchInfo) {
+	status := f.pool.TrySubmit(func() {
+		f.worker(fetchInfo.AccountFloorFetch)
+	})
+	if !status {
+		heap.Push(&f.fetchQueue, fetchInfo)
+	}
+}
+
+func (f *PriceFloorFetcher) Fetcher() {
+	//Create Ticker of 5 minutes
+	ticker := time.NewTicker(time.Duration(refetchCheckInterval) * time.Second)
+
+	for {
+		select {
+		case fetchInfo := <-f.configReceiver:
+			if fetchInfo.RefetchRequest {
+				heap.Push(&f.fetchQueue, &fetchInfo)
+			} else {
+				if _, ok := f.fetchInProgress[fetchInfo.URL]; !ok {
+					f.fetchInProgress[fetchInfo.URL] = true
+					f.submit(&fetchInfo)
+				}
+			}
+		case <-ticker.C:
+			currentTime := f.time.Now().Unix()
+			for top := f.fetchQueue.Top(); top != nil && top.FetchTime <= currentTime; top = f.fetchQueue.Top() {
+				nextFetch := heap.Pop(&f.fetchQueue)
+				f.submit(nextFetch.(*FetchInfo))
+			}
+		case <-f.done:
+			glog.Info("Price Floor fetcher terminated")
```

Comment:
Should we make the `f.Stop()` call here?

Reply:
When we close the channel, it will send single on the channel and it will take care to stop the fetcher

Reply:
Where will the `f.Stop()` call take place?

Reply:
Ideally it should be called when we are shutting down the server
Currently we are not stopping the fetcher instance, so it was not being used

Reply:
This utility is for graceful shutdown of priceFloor fetcher